processor/src/trace/virtual_bus/sum_check/verifier/mod.rs

```diff
@@ -0,0 +1,152 @@
+    pub fn new(composition_poly: P, final_query_builder: V) -> Self {
+        let max_degree = composition_poly.max_degree();
+        let eval_domain = EvaluationDomain::new(max_degree);
+
+        Self {
+            composition_poly,
+            eval_domain,
+            final_query_builder,
+            _challenger: PhantomData,
+        }
+    }
+
+    /// Verifies a sum-check proof [Proof] and returns a claim on the openings of the mult-linear
+    /// oracles that are part of the statement being proven.
+    ///
+    /// More precisely, the method:
+    ///
+    /// 1. Generates a `claimed_evaluation` from the round proof polynomials and the round challenge
+    /// randomness.
+    ///
+    /// 2. Computes a query that is built using the [FinalQueryBuilder] from the multi-linear
+    /// openings and the round challenges.
+    ///
+    /// 3. Evaluates the composition polynomial at the query and checks that it is equal
+    /// `claimed_evaluation`.
+    ///
+    /// 4. Outputs a `FinalOpeningClaim` on the multi-linear oracles.
+    ///
+    /// Thus, the proof is correct if the method outputs a [FinalOpeningClaim] and this latter is
+    /// a valid claim on the multi-linear oracles i.e., each multi-linear oracle opens to the
+    /// claimed value at the specified opening point.
+    ///
+    ///  # Errors
+    /// Returns an error if:
+    /// - No openings were provided.
+    /// - Draw the round challenge fails.
+    /// - The final evaluation check fails.
+    pub fn verify(
+        &self,
+        claim: E,
+        proof: Proof<E>,
+        coin: &mut C,
+    ) -> Result<FinalOpeningClaim<E>, Error> {
+        let Proof {
+            openings: openings_claim,
+            round_proofs,
+        } = proof;
+
+        let mut claimed_evaluation = claim;
+        let mut evaluation_point = vec![];
+        for round_proof in round_proofs {
+            let partial_evals = round_proof.poly_evals.clone();
+            coin.reseed(H::hash_elements(&partial_evals));
+            let evals = round_proof.to_evals(claimed_evaluation);
+
+            let r = coin.draw().map_err(|_| Error::FailedToGenerateChallenge)?;
+            let reduced_evaluation = self.eval_domain.evaluate(&evals, r);
+
+            claimed_evaluation = reduced_evaluation;
+            evaluation_point.push(r);
+        }
+
+        if let Some(openings_claim) = openings_claim {
+            if openings_claim.evaluation_point != evaluation_point {
+                return Err(Error::WrongOpeningPoint);
+            }
+            let query = self.final_query_builder.build_query(&openings_claim, &evaluation_point);
+            if self.composition_poly.evaluate(&query) != claimed_evaluation {
+                Err(Error::FinalEvaluationCheckFailed)
+            } else {
+                Ok(openings_claim)
+            }
+        } else {
+            Err(Error::NoOpeningsProvided)
+        }
```

Comment:
If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. *not* an `Option`)

Reply:
I just noticed the comment on `Proof`:

```rs
/// Openings is an [Option] as there are situations where we would like to run the sum-check
/// protocol for a certain number of rounds that is less than the number of variables of the
/// multi-linears. This is the case for example when we have a merged polynomial.
```

What is meant by a "merged polynomial"?

Reply:
> If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. _not_ an `Option`)

I have changed this in the subsequent PR as this was a poor design choice

Reply:
> I just noticed the comment on `Proof`:
> 
> ```rust
> /// Openings is an [Option] as there are situations where we would like to run the sum-check
> /// protocol for a certain number of rounds that is less than the number of variables of the
> /// multi-linears. This is the case for example when we have a merged polynomial.
> ```
> 
> What is meant by a "merged polynomial"?

Fixed now